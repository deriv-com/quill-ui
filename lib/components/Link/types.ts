import { TAllSizesExceptExtraSmall } from "@types";
import { LinkHTMLAttributes, ReactNode } from "react";

export type LinkSize = "caption" | TAllSizesExceptExtraSmall;
export type TLinkColor = "black" | "white";

export interface LinkProps extends LinkHTMLAttributes<HTMLAnchorElement> {
    className?: string;
    size?: LinkSize;
    color?: TLinkColor;
    disabled?: boolean;
    icon?: ReactNode;
    hasChevron?: boolean;
    target?: "_blank" | "_self" | "_parent" | "_top";
    children?: React.ReactNode;
}
