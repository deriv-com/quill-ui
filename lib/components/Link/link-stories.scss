.link-stories {
    width: 150px;
    padding: var(--core-spacing-800);
    border-radius: var(--core-borderRadius-200);
    background-color: var(--component-textIcon-normal-prominent);
}