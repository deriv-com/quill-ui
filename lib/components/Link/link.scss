.link {
    display: inline-flex;
    gap: 8px;
    text-decoration: none;
    cursor: pointer;
    align-items: center;
    &__color {
        &-black {
            color: var(--core-color-solid-slate-1400);
        }
        &-white {
            color: var(--core-color-solid-slate-50);
        }
    }
    color: var(--component-textIcon-normal-prominent);

    &:hover {
        text-decoration: underline;
    }

    &:active {
        text-decoration: underline;
    }

    &:disabled {
        pointer-events: none;
        opacity: var(--core-opacity-600);
    }

    &__icon-color {
        &-black {
            fill: var(--core-color-solid-slate-1400);
        }
        &-white {
            fill: var(--core-color-solid-slate-50);
        }
    }

    &__icon-size {
        &-caption {
            width: 8px;
            height: 18px;
        }

        &-sm {
            width: 8px;
            height: 22px;
        }

        &-md {
            width: 10px;
            height: 24px;
        }

        &-lg {
            width: 13px;
            height: 30px;
        }

        &-xl {
            width: 15px;
            height: 36px;
        }
    }

}