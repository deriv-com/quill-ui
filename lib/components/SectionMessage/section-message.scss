.section-message {
    display: flex;
    padding: var(--semantic-spacing-general-md);
    gap: var(--semantic-spacing-gap-md);
    background-color: var(--core-color-opacity-black-75);
    border-radius: var(--core-borderRadius-800);

    &--info {
        background-color: var(--core-color-opacity-blue-100);
    }
    &--success {
        background-color: var(--core-color-opacity-green-100);
    }
    &--danger {
        background-color: var(--core-color-opacity-red-100);
    }
    &--warning {
        background-color: var(--core-color-opacity-yellow-100);
    }

    &-content {
        display: flex;
        flex-direction: column;
        gap: var(--semantic-spacing-gap-lg);

        .title-and-message {
            display: flex;
            flex-direction: column;
            gap: var(--semantic-spacing-gap-md);
        }
    }

    &-links {
        display: flex;
        gap: var(--semantic-spacing-gap-lg);
    }
}
