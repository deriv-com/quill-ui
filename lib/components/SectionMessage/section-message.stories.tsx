import type { Meta, StoryObj } from "@storybook/react";
import { SectionMessage } from "./index";
import {
    StandaloneCircleUserRegularIcon,
    StandaloneAndroidIcon,
} from "@deriv/quill-icons/Standalone";

const icons = {
    sample_1: <StandaloneCircleUserRegularIcon iconSize="sm" />,
    sample_2: <StandaloneAndroidIcon iconSize="sm" />,
};

const data = [
    {
        id: 1,
        linkProps: {
            hasChevron: true,
            children: "Click here",
            href: "/",
        },
    },
    {
        id: 2,
        linkProps: {
            hasChevron: true,
            children: "Disabled Link",
            href: "/",
            disabled: true,
        },
    },
];

const links = {
    one: data.slice(0, 1),
    two: data,
    none: null,
};

const meta = {
    title: "Components/SectionMessage/SectionMessage",
    component: SectionMessage,
    argTypes: {
        size: {
            options: ["sm", "md"],
            control: "radio",
            description: "To select the size of the section-message",
        },
        status: {
            options: ["info", "success", "warning", "danger", null],
            control: { type: "radio" },
            description: "To select the status",
        },
        icon: {
            options: Object.keys(icons),
            mapping: icons,
            control: "radio",
        },
        linkList: {
            options: Object.keys(links),
            mapping: links,
            control: "radio",
        },
        className: { table: { disable: true } },
    },
    tags: ["autodocs"],
} satisfies Meta<typeof SectionMessage>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
    args: {
        title: "Default",
        message: "This is a message",
        size: "md",
        icon: icons.with_icon,
    },
};

export const Info: Story = {
    args: {
        title: "Information",
        message: "This is an information message",
        size: "md",
        status: "info",
        linkList: [
            {
                id: 1,
                linkProps: {
                    hasChevron: true,
                    children: "Click here",
                    href: "/",
                },
            },
        ],
    },
};

export const Success: Story = {
    args: {
        title: "Success",
        message: "This is a success message",
        size: "md",
        status: "success",
        linkList: [
            {
                id: 1,
                linkProps: {
                    hasChevron: true,
                    children: "Click here",
                    href: "/",
                },
            },
            {
                id: 2,
                linkProps: {
                    hasChevron: true,
                    children: "Click here",
                    href: "/",
                    disabled: true,
                },
            },
        ],
    },
};

export const Danger: Story = {
    args: {
        title: "Danger",
        message: "This is a danger message",
        size: "md",
        status: "danger",
        linkList: [
            {
                id: 1,
                linkProps: {
                    hasChevron: true,
                    children: "Click here",
                    href: "/",
                },
            },
        ],
    },
};

export const Warning: Story = {
    args: {
        title: "Warning",
        message: "This is a warning message",
        size: "md",
        status: "warning",
    },
};
