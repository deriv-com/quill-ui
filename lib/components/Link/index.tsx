import { forwardRef } from "react";
import { LabelPairedChevronRightSmRegularIcon } from "@deriv/quill-icons/LabelPaired";
import { LinkProps } from "./types";
import clsx from "clsx";
import "./link.scss";
import { CaptionText, Text } from "@components/Typography";

export const Link = forwardRef<HTMLAnchorElement, LinkProps>(
    (
        {
            className,
            size = "md",
            disabled,
            color = "black",
            icon,
            hasHoverEffect = true,
            hasChevron = false,
            children,
            target = "_blank",
            href,
            ...rest
        },
        ref,
    ) => {
        return (
            <a
                ref={ref}
                target={target}
                href={disabled ? "javascript:void(0)" : href}
                {...rest}
                className={clsx(
                    "quill-link",
                    `quill-link__color-${color}`,
                    disabled && "quill-link--disabled",
                    hasHoverEffect && "quill-link--hover",
                    className,
                )}
            >
                {icon && (
                    <span data-testid="dt-link-icon" className="quill-link__icon">
                        {icon}
                    </span>
                )}
                {size === "caption" ? (
                    <CaptionText color={color}>{children}</CaptionText>
                ) : (
                    <Text
                        color={color}
                        size={
                            size === "sm"
                                ? "sm"
                                : size === "lg"
                                  ? "lg"
                                  : size === "xl"
                                    ? "xl"
                                    : "md"
                        }
                    >
                        {children}
                    </Text>
                )}

                {hasChevron && (
                    <LabelPairedChevronRightSmRegularIcon
                        data-testid="dt-link-chevron"
                        className={clsx(
                            `quill-link__icon-color-${color}`,
                            `quill-link__icon-size-${size}`,
                        )}
                    />
                )}
            </a>
        );
    },
);

Link.displayName = "Link";

export default Link;
