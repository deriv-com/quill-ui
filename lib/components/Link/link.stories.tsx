import type { Meta, StoryObj } from "@storybook/react";
import Link from ".";
import {
    LabelPairedPlaceholderCaptionRegularIcon,
    LabelPairedPlaceholderLgRegularIcon,
    LabelPairedPlaceholderMdRegularIcon,
    LabelPairedPlaceholderSmRegularIcon,
    LabelPairedPlaceholderXlRegularIcon,
} from "@deriv/quill-icons/LabelPaired";
import { ComponentProps } from "react";

const meta = {
    title: "Components/Links/Link",
    component: Link,
    parameters: {
        layout: "centered",
    },
    args: {
        size: "md",
    },
    argTypes: {
        size: {
            options: ["caption", "sm", "md", "lg", "xl"],
            control: { type: "radio" },
            description: "To select the size of the link",
        },
        color: {
            options: ["black", "white"],
            control: { type: "radio" },
            description: "To select the color of the link",
        },
    },
    tags: ["autodocs"],
} satisfies Meta<typeof Link>;

export default meta;
type Story = StoryObj<typeof meta>;
const placeholder = {
    caption: <LabelPairedPlaceholderCaptionRegularIcon />,
    sm: <LabelPairedPlaceholderSmRegularIcon />,
    md: <LabelPairedPlaceholderMdRegularIcon />,
    lg: <LabelPairedPlaceholderLgRegularIcon />,
    xl: <LabelPairedPlaceholderXlRegularIcon />,
};
export const DefaultBlackLink: Story = {
    args: {
        children: "link",
        size: "md",
        disabled: false,
        hasChevron: true,
        color: "black",
    },
};
export const DefaultWhiteLink: Story = {
    args: {
        children: "link",
        size: "md",
        color: "white",
        disabled: false,
        hasChevron: true,
    },
};

export const LinkWithIconBlack = (args: ComponentProps<typeof Link>) => (
    <Link
        {...args}
        children="link"
        color="black"
        icon={placeholder[args?.size as keyof typeof placeholder]}
    />
);
export const LinkWithIconWhite = (args: ComponentProps<typeof Link>) => (
    <Link
        {...args}
        children="link"
        size="md"
        color="white"
        icon={<LabelPairedPlaceholderMdRegularIcon fill="#ffffff" />}
    />
);
