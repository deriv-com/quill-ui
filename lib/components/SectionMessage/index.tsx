import { ComponentProps, ReactNode } from "react";
import { TMediumSizes } from "@types";
import "./section-message.scss";
import clsx from "clsx";
import {
    StandaloneCircleCheckBoldIcon,
    StandaloneCircleExclamationBoldIcon,
    StandaloneCircleInfoBoldIcon,
    StandaloneTriangleExclamationBoldIcon,
} from "@deriv/quill-icons/Standalone";
import { Text } from "@components/Typography";
import Link from "@components/Link";
import { LinkProps } from "@components/Link/types";

export type LinkList = {
    id: number;
    linkProps: LinkProps;
};
export interface SectionMessageProps
    extends Omit<ComponentProps<"div">, "title"> {
    title?: string;
    className?: string;
    message: ReactNode;
    size: TMediumSizes;
    status?: "info" | "success" | "danger" | "warning";
    icon?: ReactNode;
    linkList?: LinkList[];
}

const Icons = {
    info: (
        <StandaloneCircleInfoBoldIcon
            fill="var(--core-color-solid-blue-800)"
            iconSize="sm"
        />
    ),
    danger: (
        <StandaloneTriangleExclamationBoldIcon
            fill="var(--core-color-solid-red-800)"
            iconSize="sm"
        />
    ),
    success: (
        <StandaloneCircleCheckBoldIcon
            fill="var(--core-color-solid-green-800)"
            iconSize="sm"
        />
    ),
    warning: (
        <StandaloneCircleExclamationBoldIcon
            fill="var(--core-color-solid-yellow-800)"
            iconSize="sm"
        />
    ),
};

export const SectionMessage = ({
    title,
    message,
    className,
    status,
    icon,
    size = "md",
    linkList,
}: SectionMessageProps) => {
    return (
        <div
            className={clsx(
                "section-message",
                `section-message--${status}`,
                className,
            )}
        >
            {(status || icon) && (
                <div>{status ? Icons?.[status] : icon && icon}</div>
            )}
            <div className="section-message-content">
                <div className="title-and-message">
                    {title && (
                        <Text bold size={size === "md" ? "lg" : "md"}>
                            {title}
                        </Text>
                    )}
                    <Text size={size}>{message}</Text>
                </div>
                {linkList && linkList.length > 0 && (
                    <div className="section-message-links">
                        {linkList.map(({ id, linkProps }) => (
                            <Link key={id} {...linkProps} size={size}>
                                {linkProps.children}
                            </Link>
                        ))}
                    </div>
                )}
            </div>
        </div>
    );
};

export default SectionMessage;
