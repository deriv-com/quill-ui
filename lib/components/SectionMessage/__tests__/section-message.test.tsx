import { render, screen } from "@testing-library/react";
import SectionMessage, { SectionMessageProps } from "..";
import { StandaloneCircleInfoBoldIcon } from "@deriv/quill-icons/Standalone";
import Link from "@components/Link";

describe("SectionMessage", () => {
    const defaultProps: SectionMessageProps = {
        title: "Test Title",
        message: "Test message content",
        size: "md",
    };

    const renderComponent = (props: Partial<SectionMessageProps> = {}) => {
        return render(<SectionMessage {...defaultProps} {...props} />);
    };

    it("renders correctly with default props", () => {
        const { container } = renderComponent();
        expect(container).toMatchSnapshot();
    });

    it("renders with an info status", () => {
        renderComponent({ status: "info" });
        expect(screen.getByText("Test Title")).toBeInTheDocument();
        expect(screen.getByText("Test message content")).toBeInTheDocument();
        expect(screen.getByRole("img", { hidden: true })).toBeInTheDocument(); // Assuming the icon is rendered with role 'img'
    });

    it("renders with a custom icon", () => {
        const customIcon = (
            <StandaloneCircleInfoBoldIcon fill="#0000FF" iconSize="sm" />
        );
        renderComponent({ icon: customIcon });
        expect(screen.getByText("Test Title")).toBeInTheDocument();
        expect(screen.getByText("Test message content")).toBeInTheDocument();
        expect(screen.getByRole("img", { hidden: true })).toBeInTheDocument(); // Assuming the icon is rendered with role 'img'
    });

    it("renders with links", () => {
        const links = [
            <Link>Example Link 1</Link>,
            <Link>Example Link 2</Link>,
        ];
        renderComponent({ links });
        expect(screen.getByText("Test Title")).toBeInTheDocument();
        expect(screen.getByText("Test message content")).toBeInTheDocument();
        expect(screen.getByText("Example Link 1")).toBeInTheDocument();
        expect(screen.getByText("Example Link 2")).toBeInTheDocument();
    });
});
