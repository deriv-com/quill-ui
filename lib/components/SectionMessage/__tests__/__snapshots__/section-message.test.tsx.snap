// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionMessage renders correctly with default props 1`] = `
<div>
  <div
    class="section-message section-message--undefined"
  >
    <div
      class="section-message-content"
    >
      <div
        class="title-and-message"
      >
        <p
          class="quill-typography quill-typography__body-text__size--lg__weight--bold__decoration--default quill-typography__color--default"
        >
          Test Title
        </p>
        <p
          class="quill-typography quill-typography__body-text__size--md__weight--regular__decoration--default quill-typography__color--default"
        >
          Test message content
        </p>
      </div>
    </div>
  </div>
</div>
`;
